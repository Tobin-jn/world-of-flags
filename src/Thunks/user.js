import { getUser, addUser } from '../utilities/API'
import { contentStatus, successfulLogin } from '../actions/userActions'

export const fetchUser = (name, email, password) => {
<<<<<<< HEAD
=======
  console.log(name, email, password)
>>>>>>> Add styling for Account and login placeholders
  return async dispatch => {
    dispatch(contentStatus('loading'))
    try {
      let response
      if (!name) {
        response = await getUser(email, password)
      } else {
        response = await addUser(name, email, password)
      }
      dispatch(successfulLogin(response))
      dispatch(contentStatus('resolved'))
    } catch (error) {
      dispatch(contentStatus(`Email & password don't match`))
    }
  }
}
