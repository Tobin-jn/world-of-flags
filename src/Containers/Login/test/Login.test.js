import React from 'react';
import { shallow, mount } from 'enzyme';
import { Login, mapStateToProps, mapDispatchToProps } from '../index';
import { fetchUser } from '../../../thunks/user'

describe('Login', () => {
  let alex;
  let wrapper;
  let mockEvent;
  let mockFunc;
  let mockUser;
  let expected;
  
  beforeEach(() => {
    mockFunc = jest.fn();
    mockEvent = {
      preventDefault: mockFunc,
      target: {
        name: 'name',
        value: 'Alex'
      }
    }

    mockUser = {
                id: 2,
                name: 'rtg',
                loggedIn: true,
                email: 'alex@turing.com'
               }

    wrapper = shallow(<Login
                        fetchUser={jest.fn()}
                        loading={'true'}
                        user={mockUser}
                      />);
  });
  
    it('should render like the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

  describe('createUser', () => {
    it('should should set state upon invocaton of createUser', () => {
      wrapper.setState({ createUser: false })
      wrapper.instance().createUser()
      expect(wrapper.state().createUser).toEqual(true)
    });
  });
  
  describe('loginUser', () => {
    it('should call fetchUser with the correct parameters if createUser is truthy', () => {
      wrapper.setState({ 
                          createUser: true,
                          name: 'Alex',
                          email: 'alex@123.com',
                          password: '123' 
                        })
      wrapper.instance().loginUser(mockEvent)
      expect(wrapper.instance().props.fetchUser).toHaveBeenCalledWith(null, 'alex@123.com', '123')
    });

    it('should call fetchUser with the correct parameters if createUser is falsy', () => {
      wrapper.setState({ 
                          createUser: false,
                          name: 'Alex',
                          email: 'alex@123.com',
                          password: '123' 
                        })
      wrapper.instance().loginUser(mockEvent)
      expect(wrapper.instance().props.fetchUser).toHaveBeenCalledWith('Alex', 'alex@123.com', '123')
    });

  });

  describe('handleChange', () => {
    it('should set state upon invocaton of handleChange', () => {
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state().name).toEqual('Alex')
    });
  });

  describe('expandCredentials', () => {
    it('should should set state upon invocaton of expandCredentials', () => {
      wrapper.setState({ emailCredentials: false })
      wrapper.instance().expandCredentials()
      expect(wrapper.state().emailCredentials).toEqual(true)
    });
  });

  describe('closeCredentials', () => {
    it('should should set state upon invocaton of closeCredentials', () => {
      wrapper.setState({ emailCredentials: true })
      wrapper.instance().closeCredentials()
      expect(wrapper.state().emailCredentials).toEqual(false)
    });
  });
  
  describe('changeFormPurpose', () => {
    it('should should set state upon invocaton of changeFormPurpose', () => {
      wrapper.setState({ formLogin: true })
      wrapper.instance().changeFormPurpose()
      expect(wrapper.state().formLogin).toEqual(false)
    });
  });

  describe('handleSubmit', () => {
    it('should invoke loginUser if loginAttempt is successful', async () => {
      wrapper.instance().loginUser = jest.fn()
      await wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.instance().loginUser).toHaveBeenCalled();
    });

    it('should return a user if component is mounted', () => {
      wrapper.setState({ _isMounted: true })
      const submit = wrapper.instance().handleSubmit(mockEvent)
      expect(submit).toBeDefined()
    })
  });
  
  describe('clearInputs', () => {
    it('should clear inputs on submit', () => {
      wrapper.setState({
        name: 'Bruce',
        email: 'Ela',
        password: 'Rau',
      });
      wrapper.instance().clearInputs();
      expect(wrapper.state().name).toEqual('');
      expect(wrapper.state().email).toEqual('');
      expect(wrapper.state().password).toEqual('');
    });
  });

  describe('mapStateToProps', () => {
    it('should create the correct props object', () => {
      let mockState = {
        user: {id: 2, user: 'Bob', email: "bob@bob.com", loggedIn: true},
        loading: 'resolved'
      }
      const  expected = {
        user: {id: 2, user: 'Bob', email: 'bob@bob.com', loggedIn: true},
        loading: 'resolved'
      }

      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    });
  });
  
  describe('mapDispatchToProps', () => {
    const mockDispatch = jest.fn()

    it('should map a key of fetchUser', () => {
      const dispatchedProps = mapDispatchToProps(mockDispatch)
      expect(dispatchedProps.fetchUser).toBeDefined()
    });
    
    it('should have fetchUser call dispatch', () => {
      const mockUser = { name: 'Alex', email: 'alex@123.com', password: '123'}

      const dispatchedProps = mapDispatchToProps(mockDispatch)
      dispatchedProps.fetchUser(mockUser)

      expect(mockDispatch).toHaveBeenCalled();
    });
  });
});