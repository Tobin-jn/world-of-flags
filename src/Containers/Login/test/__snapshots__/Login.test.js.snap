// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render like the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="login-title"
        >
          FlagShip
        </h1>,
        <div
          className="button login-button-facebook"
        >
          Facebook
        </div>,
        <div
          className="button login-button-twitter"
        >
          Twitter
        </div>,
        <div
          className="button login-button-email"
        >
          Email
        </div>,
      ],
      "className": "Login",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "FlagShip",
          "className": "login-title",
        },
        "ref": null,
        "rendered": "FlagShip",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Facebook",
          "className": "button login-button-facebook",
        },
        "ref": null,
        "rendered": "Facebook",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Twitter",
          "className": "button login-button-twitter",
        },
        "ref": null,
        "rendered": "Twitter",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Email",
          "className": "button login-button-email",
        },
        "ref": null,
        "rendered": "Email",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="login-title"
          >
            FlagShip
          </h1>,
          <div
            className="button login-button-facebook"
          >
            Facebook
          </div>,
          <div
            className="button login-button-twitter"
          >
            Twitter
          </div>,
          <div
            className="button login-button-email"
          >
            Email
          </div>,
        ],
        "className": "Login",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "FlagShip",
            "className": "login-title",
          },
          "ref": null,
          "rendered": "FlagShip",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Facebook",
            "className": "button login-button-facebook",
          },
          "ref": null,
          "rendered": "Facebook",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Twitter",
            "className": "button login-button-twitter",
          },
          "ref": null,
          "rendered": "Twitter",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Email",
            "className": "button login-button-email",
          },
          "ref": null,
          "rendered": "Email",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
